slow_conv2d grad_input: avoid dispatch in parallel region

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65759
* #65758
* #65757
* #65737
* #65726
* __->__ #65725
* #65724

See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Call `grad_input.zero_()` only once, outside of the parallel region
3. Replace `at::mm` with a `gemm` call

Differential Revision: [D31257876](https://our.internmc.facebook.com/intern/diff/D31257876)

peterbell10 committed:
slow_conv2d grad_input: avoid dispatch in parallel region

See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Call `grad_input.zero_()` only once, outside of the parallel region
3. Replace `at::mm` with a `gemm` call

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d grad_input: avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Call `grad_input.zero_()` only once, outside of the parallel region
3. Replace `at::mm` with a `gemm` call

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d grad_input: avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Call `grad_input.zero_()` only once, outside of the parallel region
3. Replace `at::mm` with a `gemm` call

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -270,27 +286,35 @@ void slow_conv2d_backward_out_cpu_template(
+    auto weight_a = weight.accessor<scalar_t, 2>();
+
+    at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
+      auto fgrad_input = std::make_unique<scalar_t[]>(
```
nice allocation

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257876).

peterbell10 committed:
Update on "slow_conv2d grad_input: avoid dispatch in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Call `grad_input.zero_()` only once, outside of the parallel region
3. Replace `at::mm` with a `gemm` call

Differential Revision: [D31257876](https://our.internmc.facebook.com/intern/diff/D31257876)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257876).
